File: frontend/src/api/monitor.ts
import request from './index'

export interface ConnectionInfo {
  id: string
  vehicle_id: string
  client_ip: string
  client_port: number
  server_ip: string
  server_port: number
  status: 'connected' | 'disconnected' | 'error'
  connect_time: string
  last_heartbeat: string
  data_count: number
  error_count: number
}

export interface SystemStatus {
  cpu_usage: number
  memory_usage: number
  disk_usage: number
  network_in: number
  network_out: number
  active_connections: number
  total_connections: number
  uptime: number
}

export interface LogEntry {
  id: number
  level: 'DEBUG' | 'INFO' | 'WARNING' | 'ERROR'
  message: string
  timestamp: string
  source: string
  details?: string
}

export interface MonitorQuery {
  level?: string
  start_time?: string
  end_time?: string
  page?: number
  size?: number
}

// 获取连接信息列表
export const getConnections = () => {
  return request.get<ConnectionInfo[]>('/monitor/connections')
}

// 获取连接详情
export const getConnection = (id: string) => {
  return request.get<ConnectionInfo>(`/monitor/connections/${id}`)
}

// 断开连接
export const disconnectConnection = (id: string) => {
  return request.delete(`/monitor/connections/${id}`)
}

// 获取系统状态
export const getSystemStatus = () => {
  return request.get<SystemStatus>('/monitor/system')
}

// 获取系统状态历史
export const getSystemStatusHistory = (hours: number = 24) => {
  return request.get<SystemStatus[]>('/monitor/system/history', {
    params: { hours }
  })
}

// 获取日志列表
export const getLogs = (params: MonitorQuery) => {
  return request.get<LogEntry[]>('/monitor/logs', { params })
}

// 获取日志详情
export const getLog = (id: number) => {
  return request.get<LogEntry>(`/monitor/logs/${id}`)
}

// 清除日志
export const clearLogs = (before?: string) => {
  return request.delete('/monitor/logs', {
    params: { before }
  })
}

// 获取实时监控数据
export const getRealTimeData = () => {
  return request.get('/monitor/realtime')
}

// 获取性能统计
export const getPerformanceStats = (period: 'hour' | 'day' | 'week' = 'day') => {
  return request.get('/monitor/performance', {
    params: { period }
  })
}

// 获取告警列表
export const getAlerts = (params: MonitorQuery) => {
  return request.get('/monitor/alerts', { params })
}

// 确认告警
export const acknowledgeAlert = (id: number) => {
  return request.put(`/monitor/alerts/${id}/acknowledge`)
}

// 获取监控配置
export const getMonitorConfig = () => {
  return request.get('/monitor/config')
}

// 更新监控配置
export const updateMonitorConfig = (config: any) => {
  return request.put('/monitor/config', config)
} 

File: api/routers/monitor.py
from fastapi import APIRouter, HTTPException, Query
from typing import List, Optional
from datetime import datetime, timedelta
import psutil
import os
from ..services.monitor_service import MonitorService

router = APIRouter(prefix="/monitor", tags=["监控"])

# 创建监控服务实例
monitor_service = MonitorService()

@router.get("/connections")
async def get_connections():
    """获取连接列表"""
    try:
        connections = monitor_service.get_connections()
        return {
            "code": 200,
            "message": "获取连接列表成功",
            "data": connections
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"获取连接列表失败: {str(e)}")

@router.get("/system")
async def get_system_status():
    """获取系统状态"""
    try:
        status = monitor_service.get_system_status()
        return {
            "code": 200,
            "message": "获取系统状态成功",
            "data": status
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"获取系统状态失败: {str(e)}")

@router.get("/realtime")
async def get_real_time_data():
    """获取实时监控数据"""
    try:
        data = monitor_service.get_real_time_data()
        return {
            "code": 200,
            "message": "获取实时数据成功",
            "data": data
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"获取实时数据失败: {str(e)}")

@router.get("/performance")
async def get_performance_stats(
    period: str = Query("day", description="统计周期: hour, day, week")
):
    """获取性能统计"""
    try:
        stats = monitor_service.get_performance_stats(period)
        return {
            "code": 200,
            "message": "获取性能统计成功",
            "data": stats
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"获取性能统计失败: {str(e)}")

@router.delete("/connections/{connection_id}")
async def disconnect_connection(connection_id: str):
    """断开指定连接"""
    try:
        success = monitor_service.disconnect_connection(connection_id)
        if success:
            return {
                "code": 200,
                "message": "断开连接成功",
                "data": None
            }
        else:
            raise HTTPException(status_code=404, detail="连接不存在")
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"断开连接失败: {str(e)}") 